master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,667 @@
+		} else {
+			instancesToTerminate = append(instancesToTerminate, instanceId)
+		}
+	}
+
+	ctx.Log().Debugf(
+		"terminating %d EC2 instances and %d spot requests: %s,  %s",
+		len(instancesToTerminate),
+		len(pendingSpotRequestsToTerminate),
+		instancesToTerminate,
+		pendingSpotRequestsToTerminate,
+	)
+
+	terminateInstancesResponse, err := c.terminateInstances(instancesToTerminate, false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate EC2 instances")
+	} else {
+		terminated := c.newInstancesFromTerminateInstancesOutput(terminateInstancesResponse)
+		ctx.Log().Debugf(
+			"terminated %d EC2 instances: %s",
+			len(terminated),
+			fmtInstances(terminated),
+		)
+	}
+
+	_, err = c.terminateSpotInstanceRequests(ctx, pendingSpotRequestsToTerminate, false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate spot requests")
+	} else {
+		ctx.Log().Debugf(
+			"terminated %d spot requests: %s",
+			len(pendingSpotRequestsToTerminate),
+			pendingSpotRequestsToTerminate,
+		)
+	}
+
+	// TODO: Race condition - an instance could have been created between listing
+	//       and terminating spot request. We could clean up the spot instances here.
+	//       But it will be cleaned up on the next call to list() anyway.
+}
+
+func (c *awsCluster) launchSpot(
+	ctx *actor.Context,
+	instanceType instanceType,
+	instanceNum int,
+) {
+	if c.spotLoopState.inBackoffState {
+		backoffEnd := c.spotLoopState.backoffStart.Add(backoffDuration)
+		if time.Now().Before(backoffEnd) {
+			ctx.Log().Infof("AWS spot provider refusing to launch spot. spot provider is in backoff " +
+				"state because all recent spot requests have failed")
+			return
+		}
+	}
+
+	instType, ok := instanceType.(ec2InstanceType)
+	if !ok {
+		panic("cannot pass non-ec2InstanceType to ec2Cluster")
+	}
+
+	if instanceNum < 0 {
+		return
+	}
+
+	ctx.Log().Infof("launching %d EC2 spot requests", instanceNum)
+	resp, err := c.createSpotInstanceRequestCorrectingForClockSkew(ctx, instanceNum, false, instType)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot launch EC2 spot requests")
+		return
+	}
+
+	// Update the internal spotRequest tracker because there can be a large delay
+	// before the API starts including these requests in listSpotRequest API calls,
+	// and if we don't track it internally, we will end up overprovisioning.
+	for _, request := range resp.SpotInstanceRequests {
+		c.spotLoopState.activeSpotRequests[*request.SpotInstanceRequestId] = &spotRequest{
+			SpotRequestId: *request.SpotInstanceRequestId,
+			State:         *request.State,
+			StatusCode:    request.Status.Code,
+			StatusMessage: request.Status.Message,
+			CreationTime:  request.CreateTime,
+			InstanceId:    nil,
+		}
+
+		ctx.Log().Infof(
+			"Launching spot request, %s, %s",
+			*request.SpotInstanceRequestId,
+			*request.State,
+		)
+	}
+	return
+
+}
+
+func (c *awsCluster) attemptToApproximateClockSkew(ctx *actor.Context) {
```

Comment:
I think we need a high-level explanation of what this function does and why we need this